### ISSUE TITLE
Time-dependent functions don't work properly

### ISSUE THREAD

**Issue by ichnilatis-gr** [koreader/koreader#9493]
* KOReader version: 2022.8
* Device: PB 628

#### Issue
I [have written](https://www.mobileread.com/forums/showthread.php?t=347764) in "mobile read" forum that I had noticed that, when the battery level is lower than 50%, every function that depends on the time (e.g. auto warmth, the clock in the status bar etc.) doesn't work properly. After the last update, I noticed that it happens even when the battery is full or almost full.

#### Steps to reproduce
For example, if I enable the "autoturn", this doesn't work. Also, in "auto warmth" mode, the night mode takes effect when I close the book.

Ηave you noticed the same problem?

**Comment by ichnilatis-gr** [comment]
Well, I'll give you some examples:

- If I've set the night mode to take effect at a certain time and I'm reading a book at that time, the night mode will not take effect until I close the book.
- The "autoturn" doesn't work at all.
- The time in the status bar doesn't change until I turn the page.

I had noticed that these problems were happening when the battery was lower than 50-60%, but now they happen even when the battery is at a high level.
Note that I use my device offline. Once that I used it online, I noticed that there were no such problems, even if the battery was at a low level.

I don't know if I can help you otherwise.
Thanks.

**Comment by NiLuJe** [comment]
It appears to be standby related, so only someone intimately familiar with how fucked up it might be on PB can help. That probably currently amounts to a neat 0 in the pool of active contributors, hence the tags ;).

**Comment by pazos** [comment]
I assumed than disabling "autostanby" plugin is going to disable standby on pocketbooks. I might to be wrong.

Things you can try:

1. Enable "autostandby plugin"
2. Restart KOReader
3. Go to "gear" -> "device" -> "autostandby settings" and tune the settings:

see if you can check "forbidden". This is going to eat your battery and disable autostandby completely. Test the behaviour. If all works as expected you won't see any issue other than your battery reduced to a 10-30% based on your usage patterns.

If that works as intended and you want to get a better battery behaviour disable "forbidden" again and just play with the other settings.  They're documented in https://github.com/koreader/koreader/blob/master/plugins/autostandby.koplugin/main.lua#L32-L38.

Keep us informed :)

**Comment by NiLuJe** [comment]
AFAIU, the two are separate (i.e., there's a dedicated system settings we don't touch). Whether the plugin actually *requires* the system setting to be enabled to actually do something is something I'm not sure of, though.

**Comment by ichnilatis-gr** [comment]
> Keep us informed :)

First I followed the steps described when the battery level was low. It didn't work.
I have to say that there isn't the option "forbidden", unless you mean to disable autostandby by unchecking it.
Then, after I charged the device, the time-dependent functions only worked when I had autostandby unchecked. When I disabled the "autostandby plugin", they didn't work.

I wonder how this plugin affects these functions...

**Comment by pazos** [comment]
> Then, after I charged the device, the time-dependent functions only worked when I had autostandby unchecked. When I disabled the "autostandby plugin", they didn't work.

The implied solution here: enable autostandby plugin **and**, in autostandby plugin settings, disable autostandby. Spend a few days with the new config to gather info and report back.


> I wonder how this plugin affects these functions...

The plugin interacts with a core feature on pocketbooks. It seems I was wrong and disabling the plugin doesn't turn off the core feature. For that you'll need to do the steps above. 

If you still experience bad behaviour with this setup then I'm afraid there's nothing to do. Of course bad behaviour here excludes battery drain, which can be workarounded by changing the other settings while keep autostandby enabled. But **first** try the hard way.

**Comment by ichnilatis-gr** [comment]
> Of course bad behaviour here excludes battery drain, which can be workarounded by changing the other settings while keep autostandby enabled.

If I activate the autostandby, the "time-dependent functions" don't work.
I also noticed that the battery statistics do not keep actual records, when the "autostandby plugin" is enabled. Once I disabled the plugin, I saw that the battery statistics were counting the time since I deactivated the plugin and, of course the percentage of the battery consumed wasn't the real one.

**Comment by NiLuJe** [comment]
The Battery Statistics plugin is a broken mess that doesn't deal with a lot of things properly, it's by no means a good benchmark of sane behavior ;).

**Comment by pazos** [comment]
> If I activate the autostandby, the "time-dependent functions" don't work.

They aren't going to work, **ever**.

But you can enlarge the window frame where touching the screen increases the time the cpu is is alive. To a point where some of the oddities you find are no longer there. But first spend a long time with autostandby disabled to check that all behave properly. Weeks better than days, days better than hours.

No other user is reporting issues like you and **there're issues**, so it means that nobody cares enough. You're the entry point of feedback in the pocketbook world.

**Comment by rjd22** [comment]
I've been looking at this issue and can confirm that all time dependent functions seem to break when auto standby is enabled until the device is activated again by touch.

The auto standby feature puts the pocketbook in a really deep sleep saving a lot of power but as a side effect also stops all non-native timer functions.

I'm not sure if it will be easy to fix this but it would require scheduling all timed functions trough the native Inkview timer or event handling functions. And event then I'm not sure if the PocketBook will run them when being in deep sleep.

https://github.com/ezdiy/arm-pocketbook-linux-gnueabi/blob/master/sysroot/usr/include/inkview.h#L1776:L1798

I've not been around much because of other projects, so I didn't have time to look deeper into this yet.

**Comment by NiLuJe** [comment]
> Edit: Aren't we confusing auto-standby and auto-suspend?

AutoSuspend is our (mostly) platform-independent plugin for, well, auto suspend & autoshutdown.

Then came AutoStandby, which is a platform-specific plugin for PB that handles auto-standby (i.e., suspend-with-lights-and-input-on), which ties into the native facilities for that on PocketBook, as well as adds a few facilities in KOReader's main loop to deal with it.

Then, @zwim introduced a less platform-specific implementation of that for use on Kobo, where the system also natively does that, but where we don't have any middleware to go through. This was implemented completely inside the AutoSuspend plugin.

The common denominator between the devices where this is helpful is that they all run on crapp-ass sunxi SoCs ;).

**Comment by NiLuJe** [comment]
TL;DR: How this is handled in the non-PB codepath is that we setup an rtc wake right *after* our next scheduled task, which allows it to run on resume.

I have no idea how the PB stuff works, and if it even accounts for that, or if it even can.

(But that'd be a "hell no" on tying more InkView shenanigans in the main loop, FWIW).

**Comment by pazos** [comment]
I must insist: tweaking the time window is enough for most usages. Specific plugins that require the device alive (like the one that does auto page forward, autopageturn?) might be better if they requested that to the autostandby plugin via events/dispatcher.

**Comment by rjd22** [comment]
@NiLuJe we might be able to solve this problem in that case easily by disabling auto-standby in the PocketBook code and implementing the Device:standby method calling the following Inkview method:

 https://github.com/ezdiy/arm-pocketbook-linux-gnueabi/blob/master/sysroot/usr/include/inkview.h#L2381

I haven't tested it yet but it might be that the ms parameter is the amount of milliseconds that the device should sleep before the next task.

**Comment by NiLuJe** [comment]
Hmm, depends on whether this is suspend or standby... (the `deep` argument is not necessarily a good sign, for instance ;)).

**Comment by ichnilatis-gr** [comment]
> You're the entry point of feedback in the pocketbook world.

As you can see, there are other persons also in the "pocketbook world", who know better how to point out and describe the problems. 🙂 

Thank you @rjd22 for your help!

**Comment by rjd22** [comment]
I"ve been working on this today and are able to get the PocketBook to sleep and wake up by using the GoSleep method and disabling Inkview autoSuspend and the autostandby plugin.

The odd thing is that I see the device go to sleep and waking up after a minute in the logs like I want it to but the auto page turnover doesn't get rescheduled.

Might someone know why this could be?

**Comment by rjd22** [comment]
@NiLuJe @Frenzie in that case would it be preferred if I make an PR with pocketbook support for the auto-standby plugin so we can remove the auto-suspend plugin from the KoReader repo?

This would require quite some nightly testing but would bring the development of the PocketBook device closer to other vendors.

**Comment by Frenzie** [comment]
Sounds good to me provided it doesn't make AutoStandby into a convoluted mess.

**Comment by rjd22** [comment]
@Frenzie actually no changes seem to be needed to auto-standby since it supports calls to device methods. I will make a PR in a while and start testing it.

**Comment by ichnilatis-gr** [comment]
I keep noticing that the problem occurs when the battery is below 60% and while reading a book. Not while the File Manager or similar windows are open. Also, while it affects the night mode it doesn't affect the warm light, when "auto warmth" mode is activated.

**Comment by ichnilatis-gr** [comment]
> > I keep noticing that the problem occurs when the battery is below 60%
> 
> That's expected, see https://github.com/koreader/koreader/blob/master/plugins/autostandby.koplugin/main.lua#L38
> 
> You can tweak those values under "settings-> device"

But I have not the "auto-standby" activated...

**Comment by pazos** [comment]
> > > I keep noticing that the problem occurs when the battery is below 60%
> > 
> > 
> > That's expected, see https://github.com/koreader/koreader/blob/master/plugins/autostandby.koplugin/main.lua#L38
> > You can tweak those values under "settings-> device"
> 
> But I have not the "auto-standby" activated...

Yep, that's one of the problems: https://github.com/koreader/koreader/issues/9493#issuecomment-1251535780

**Comment by ichnilatis-gr** [comment]
Indeed, this seems to be the solution: I enabled autostandby and set a lower value for "Always standby if battery below" (e.g. 10%). Then I disabled autostandby (not the plugin itself). The "time-dependent functions" work properly.

Thank you!

**Comment by zmixnyc** [comment]
> Great!
> 
> I'm reopening the ticket, since it contains useful info, and close a few others that are related.

https://github.com/koreader/koreader/issues/10058#issuecomment-2046093485

**Comment by ichnilatis-gr** [comment]
> @NiLuJe @Frenzie in that case would it be preferred if I make an PR with pocketbook support for the auto-standby plugin so we can remove the auto-suspend plugin from the KoReader repo?
> 
> This would require quite some nightly testing but would bring the development of the PocketBook device closer to other vendors.

Dear @rjd22 is there anything new on this? Is there any chance that this problem will be resolved? there is a [new issue](#12323) that has emerged.
Thank you!

**Comment by rjd22** [comment]
@ichnilatis-gr I never took the time to finish it since it didn't really solve any big problems. I will have a look at my repo's to see if I can gather up an PR.

**Comment by sebastien-maillard** [comment]
Hi @rjd22 , I recently acquired a PB InkPad 3 and it seems I'm running into some of the [standby shenanigans](https://github.com/koreader/koreader/discussions/12791#discussioncomment-11363725). I have no experience with the PB platform or the KOReader code-base, but let me know if I can be of any help, maybe by testing some of your updates. I'd be happy to offer more help, but the learning curve for this topic seems pretty steep!

My issue with the _Auto Standby_ plugin is that my device never seems to enter the standby mode as long as the battery level is above the _Always standby if battery below_ threshold. 

[I was actually hoping](https://github.com/koreader/koreader/discussions/12791#discussion-7541759) to enable the `AutoSuspend` plugin for my device because I would like it to enter sleep mode (a.k.a. suspend a.k.a. screen lock a.k.a. "Zzzz") after a long period of inactivity. I haven't been able to enable the plugin so far, but I see [you suggested](#issuecomment-1328198502) to remove it entirely, after you were able to successfully [send the device to sleep](#issuecomment-1306143300):
> I"ve been working on this today and are able to get the PocketBook to sleep and wake up by using the GoSleep method and disabling Inkview autoSuspend and the autostandby plugin.

Are we talking about the same sleep mode here? Or are you referring to a standby state, following [the definition](https://github.com/koreader/koreader/issues/10117#issuecomment-1429708968) @pazos provided?
